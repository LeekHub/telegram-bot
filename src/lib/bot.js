import TelegramBot from 'node-telegram-bot-api'
import { botToken } from '../../config'
import { handleCallbackQuery } from '../handlers/callbackQuery'
import { handleInlineQuery } from '../handlers/inlineQuery'
const Agent = require('socks5-https-client/lib/Agent')

const bot = new TelegramBot(botToken, {
  polling: true,
  request: {
    agentClass: Agent,
    agentOptions: {
      socksHost: '127.0.0.1',
      socksPort: '1086'
    }
  }
})
bot.route = (routeConfig) => {
  Object.entries(routeConfig).forEach(([regex, handler]) => {
    bot.onText(new RegExp(regex), handler.bind(bot))
  })
}

bot.on('callback_query', handleCallbackQuery.bind(bot))
bot.on('inline_query', handleInlineQuery.bind(bot))

bot.sendLoadingMsg = async (chatId) => {
  const { message_id } = await bot.sendMessage(chatId, '处理中，请稍候...', {
    disable_notification: true
  })

  return () => {
    return bot.deleteMessage(chatId, message_id)
  }
}

bot.sendStockIdNotFoundError = (chatId, stockId) => {
  bot.sendMessage(chatId, `查无 ${stockId}，请确认此股票已上市/柜`)
}

bot.sendTimeoutError = async (chatId) => {
  const { message_id } = await bot.sendMessage(
    chatId,
    '發生了些問題，请再試一次...'
  )
  setTimeout(() => {
    bot.deleteMessage(chatId, message_id)
  }, 5000)
}

bot.on('message', (msg) => {
  const chatId = msg.chat.id
  // send a message to the chat acknowledging receipt of their message
  bot.sendMessage(chatId, 'Received your message')
})

export default bot
