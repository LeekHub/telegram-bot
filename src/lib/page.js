import puppeteer from 'puppeteer'

const newBrowserAndPage = async () => {
  const browser = await puppeteer.launch({
    headless: true,
    args: ['--no-sandbox']
  })

  const page = await browser.newPage()
  return [browser, page]
}

/**
 * Take screenshot for specific element
 * @param {String} url where to screenshot
 * @param {String} locator css selector for screenshot target
 * @param {Object} options
 * @param {(String|Number|Function)} options.waitFor a selector, predicate or timeout to wait for
 * @param {(String|String[])} options.waitUntil <"load"|"domcontentloaded"|"networkidle0"|"networkidle2"|Array> When to consider navigation succeeded
 * @returns {Buffer} screenshot base64 buffer
 */
export const screenshot = async (
  url,
  locator,
  { waitFor, waitUntil = 'load' } = {}
) => {
  const [browser, page] = await newBrowserAndPage()
  await page.goto(url, { waitUntil })
  await page.waitFor(waitFor || locator)
  const $elem = await page.$(locator)
  const elemImg = await $elem.screenshot({ encoding: 'base64' })
  browser.close()

  return Buffer.from(elemImg, 'base64')
}

/**
 * Get inner text of target element
 * @param {String} url
 * @param {String} locator target selector to get inner text
 * @param {(String|String[])} options.waitUntil <"load"|"domcontentloaded"|"networkidle0"|"networkidle2"|Array> When to consider navigation succeeded
 * @returns {String}
 */
export const getElementInnerText = async (
  url,
  locator,
  { waitUntil = 'load' } = {}
) => {
  const [browser, page] = await newBrowserAndPage()
  await page.goto(url, { waitUntil })
  await page.waitFor(locator)
  const text = await page.$eval(locator, ($elem) => $elem.textContent)
  browser.close()

  return text
}
