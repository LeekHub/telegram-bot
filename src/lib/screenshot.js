import puppeteer from 'puppeteer'

/**
 * Take screenshot for specific element
 * @param {string} url where to screenshot
 * @param {string} locator css selector for screenshot target
 * @returns {Buffer} screenshot base64 buffer
 */
const screenshot = async (url, locator) => {
  const browser = await puppeteer.launch({
    headless: true,
    args: ['--no-sandbox']
  })

  const page = await browser.newPage()
  await page.goto(url)
  const $chart = await page.$(locator)
  const chartImg = await $chart.screenshot({ encoding: 'base64' })
  await browser.close()

  return Buffer.from(chartImg, 'base64')
}

export default screenshot
