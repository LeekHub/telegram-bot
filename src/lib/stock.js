import fetch from 'isomorphic-fetch'
import cheerio from 'cheerio'
import { stockRawUrl } from '../../config'

export const isStockIdValid = (stockId) => {
  return /^[0-9A-Z]+$/.test(stockId)
}

// ====================
// stock quote raw data
// ====================
/**
 * Parse stock raw data from invalid JSON-like String to Object
 * @param {string} res response of fetch stock raw data
 * @returns {(object|null)}
 */
export const parseRaw = (res) => {
  try {
    return JSON.parse(
      res
        .match(/({.*})/)[1]
        .replace(/:\s?((?:-)?[0-9.]+)[,}]/g, (m, t) => m.replace(t, `"${t}"`))
    )
  } catch {
    return null
  }
}

/**
 * Stock data object normalized by stockDataNormalizer()
 * @typedef {object} NormalizedStockData
 * @property {string} stockId "2330"
 * @property {string} currentPrice "447"
 * @property {string} risePrice "+3"
 * @property {string} risePricePerc "+1.2%"
 * @property {string} amount "29865"
 * @property {string} sellAmount "18873"
 * @property {string} buyAmount "11010"
 * @property {string} maxPrice "455"
 * @property {string} minPrice "446.5"
 * @property {string} openPrice "453"
 * @property {string} lastPrice "447"
 * @property {string} turnover 成交量 (億)
 * @property {array} ticks [[price, amount]]
 */

/**
 * Normalize parsed data to object w/ necessary fields
 * @param {object} data  stock data parsed by parseRaw()
 * @returns {NormalizedStockData}
 */
export const stockDataNormalizer = (data) => {
  const { mem = {}, id: stockId } = data || {}
  const name = mem.name

  if (!name) {
    return {}
  }

  const currentPrice = mem[125]
  const risePrice = (mem[184] >= 0 ? '+' : '') + mem[184]
  const risePricePerc =
    (mem[185] >= 0 ? '+' : '') + (+mem[185]).toFixed(2).toString() + '%'
  const amount = mem[404]
  const sellAmount = mem[645]
  const buyAmount = mem[646]
  const maxPrice = mem[130]
  const minPrice = mem[131]
  const openPrice = mem[126]
  const lastPrice = mem[129]
  const turnover = (mem[423] / 100).toFixed(2).toString()
  let ticks = [
    [mem[101] > 100000 ? '市价' : mem[101], mem[113]],
    [mem[102], mem[114]],
    [mem[103], mem[115]],
    [mem[104], mem[116]],
    [mem[105], mem[117]],
    [mem[106], mem[118]],
    [mem[107], mem[119]],
    [mem[108], mem[120]],
    [mem[109], mem[121]],
    [mem[110], mem[122]]
  ] // [[price, amount]]

  ticks = ticks.map((tick) => {
    return [tick[0] || '', tick[1] || '']
  })

  return {
    stockId,
    name,
    currentPrice,
    risePrice,
    risePricePerc,
    amount,
    sellAmount,
    buyAmount,
    maxPrice,
    minPrice,
    openPrice,
    lastPrice,
    turnover,
    ticks
  }
}

/**
 * Fetch stock quotes
 * @param {string} stockId
 * @returns {NormalizedStockData}
 */
export const fetchStockData = async (stockId) => {
  const res = await fetch(
    stockRawUrl.replace('STOCK_ID', stockId)
  ).then((res) => res.text())
  console.log(stockRawUrl.replace('STOCK_ID', stockId))
  const parsedRes = parseRaw(res)
  const result = stockDataNormalizer(parsedRes)

  if (!result.name) {
    return {}
  }

  return result
}

// ====================
// crawl stock news data
// ====================
/**
 * News data crawled
 * @typedef {object} NewsItem
 * @property {string} title Article title
 * @property {string} link Article link
 * @property {string} source Article publish date & provider
 */

/**
 * Crawl news list from yahoo stock
 * @param {string} stockId
 * @returns {(NewsItem[]|null)}
 */
export const fetchStockNews = async (stockId) => {
  const res = await fetch(`https://tw.stock.yahoo.com/q/h?s=${stockId}`)
  if (!res || !res.ok || !/tw\.stock\.yahoo\.com\/q\/h\?s=/.test(res.url)) {
    return null
  }

  const doc = await res.text()
  const $ = cheerio.load(doc)
  const list = $('table table table table tbody tr')
    .toArray()
    .reduce((list, tr, i) => {
      if (i % 2 === 0) {
        const $link = $(tr).find('td a')
        const newsItem = {
          title: $link.text(),
          link: 'https://tw.stock.yahoo.com' + $link.attr('href'),
          source: ''
        }

        return [...list, newsItem]
      } else {
        list[list.length - 1].source = $(tr).find('font').text()
        return list
      }
    }, [])

  return list
}
