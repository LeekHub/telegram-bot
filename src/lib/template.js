export const getIndexHTMLTemplate = ({
  name,
  currentPrice = '-',
  risePrice = '+0',
  risePricePerc = '+0%',
  turnover = '0'
}) => {
  return `<b>${name}</b> ${currentPrice} | ${risePrice} (${risePricePerc})
成交金額 ${turnover} 億`
}

export const getStockHTMLTemplate = ({
  stockId,
  name,
  currentPrice = '-',
  risePrice = '+0',
  risePricePerc = '+0%',
  amount = '-',
  sellAmount = '-',
  buyAmount = '-',
  maxPrice = '-',
  minPrice = '-',
  openPrice = '-',
  lastPrice = '-',
  ticks
}) => {
  return `<pre>${stockId} ${name} | <b>${currentPrice}</b> | ${risePrice} (${risePricePerc})

總量 ${amount} | 內盤 ${sellAmount} | 外盤 ${buyAmount}

最高 ${maxPrice} | 最低 ${minPrice} | 開盤 ${openPrice} | 昨收 ${lastPrice}

        委買        |        委賣        
----------------------------------------
${ticks
  .map((tick, i) => {
    let s = ''
    if (i % 2 === 0) {
      s += (tick[1] + ' ' + tick[0].padStart(7, ' ')).padStart(19, ' ') + '  '
    } else {
      s += (tick[0].padEnd(7, ' ') + ' ' + tick[1]).padEnd(19, ' ') + '\n'
    }

    return s
  })
  .join('')}
</pre>
`
}

export const getNewsListHTMLTemplate = (list) => {
  return list
    .map(
      ({ title, link, source }) => `<a href="${link}">${title}</a>\n${source}\n`
    )
    .join('\n')
}

export const getUpDownIcon = (val) => {
  return val > 0 ? '🔼 ' : val < 0 ? '🔽 ' : ''
}

export const getStockCaptionTextTemplate = (stockData) => {
  const { stockId, name, currentPrice, risePrice, risePricePerc } = stockData

  const icon = getUpDownIcon(risePrice)

  return `${icon}${stockId} ${name} ${currentPrice} | ${risePrice} (${risePricePerc})`
}

export const getIndexCaptionTextTemplate = (stockData) => {
  const { name, currentPrice, risePrice, risePricePerc } = stockData

  const icon = getUpDownIcon(risePrice)

  return `${icon}${name} ${currentPrice} | ${risePrice} (${risePricePerc})`
}
