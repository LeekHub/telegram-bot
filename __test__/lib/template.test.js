const {
  getIndexHTMLTemplate,
  getStockHTMLTemplate,
  getNewsListHTMLTemplate,
  getUpDownIcon,
  getStockCaptionTextTemplate,
  getIndexCaptionTextTemplate
} = require('../../src/lib/template')
import { stockDataNormalizer } from '../../src/lib/stock'
import mockStockLimitUpData from '../mock/stock_limit_up.json'
import mockStockNews from '../mock/stock_news.json'
const mockStockData = stockDataNormalizer(mockStockLimitUpData)

describe('lib/template getIndexHTMLTemplate()', () => {
  it('should return correct template', () => {
    expect(getIndexHTMLTemplate(mockStockData)).toMatchSnapshot()
  })
})

describe('lib/template getStockHTMLTemplate()', () => {
  it('it should return correct template', () => {
    expect(getStockHTMLTemplate(mockStockData)).toMatchSnapshot()
  })
})

describe('lib/template getNewsListHTMLTemplate()', () => {
  it('it should return correct template', () => {
    expect(getNewsListHTMLTemplate(mockStockNews)).toMatchSnapshot()
  })
})

describe('lib/template getUpDownIcon()', () => {
  it('it should return correct icon', () => {
    expect(getUpDownIcon('1')).toEqual('ðŸ”¼ ')
    expect(getUpDownIcon('-1')).toEqual('ðŸ”½ ')
    expect(getUpDownIcon('0')).toEqual('')
  })
})

describe('lib/template getStockCaptionTextTemplate()', () => {
  it('it should return correct template', () => {
    expect(getStockCaptionTextTemplate(mockStockData)).toMatchSnapshot()
  })
})

describe('lib/template getIndexCaptionTextTemplate()', () => {
  it('it should return correct template', () => {
    expect(getIndexCaptionTextTemplate(mockStockData)).toMatchSnapshot()
  })
})
