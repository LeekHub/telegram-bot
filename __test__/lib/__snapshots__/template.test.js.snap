// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/template getIndexCaptionTextTemplate() it should return correct template 1`] = `"🔼 元晶 24.85 | +2.25 (+9.96%)"`;

exports[`lib/template getIndexHTMLTemplate() should return correct template 1`] = `
"<b>元晶</b> 24.85 | +2.25 (+9.96%)
成交金額 9.75 億"
`;

exports[`lib/template getNewsListHTMLTemplate() it should return correct template 1`] = `
"<a href=\\"https://tw.stock.yahoo.com/news/%E5%8F%B0%E7%81%A3%E5%8D%8A%E5%B0%8E%E9%AB%94%E7%94%A2%E6%A5%AD%E5%85%A8%E7%90%83%E7%AC%AC2-%E5%88%86%E6%9E%90%E5%B8%AB-%E8%81%9A%E8%90%BD%E5%AE%8C%E6%95%B4%E6%98%AF%E5%84%AA%E5%8B%A2-113821001.html\\">台灣半導體產業全球第2  分析師：聚落完整是優勢</a>
(2020/10/01 中央社)

<a href=\\"https://tw.stock.yahoo.com/news/%E6%96%B0%E5%8F%B0%E5%B9%A3%E6%80%A5%E5%8D%87-%E7%A7%91%E6%8A%80%E5%BB%A0%E6%86%82%E5%8C%AF%E6%90%8D%E8%A1%9D%E6%93%8A%E7%AC%AC3%E5%AD%A3%E7%8D%B2%E5%88%A9-095220930.html\\">新台幣急升 科技廠憂匯損衝擊第3季獲利</a>
(2020/10/01 中央社)

<a href=\\"https://tw.stock.yahoo.com/news/%E5%8F%B0%E8%82%A1%E7%9B%A4%E5%BE%8C-%E5%8F%B0%E8%82%A1%E7%A7%8B%E7%AF%80%E5%89%8D%E7%8B%B9%E5%B9%85%E9%9C%87%E7%95%B6-%E4%BB%8D%E5%9B%9E12500%E9%BB%9E%E4%B9%8B%E4%B8%8A-061635774.html\\">【台股盤後】台股秋節前狹幅震當 仍回12500點之上</a>
(2020/09/30 中央社)

<a href=\\"https://tw.stock.yahoo.com/news/%E7%AB%B9%E7%A7%91%E6%9C%AA%E5%8F%97%E5%9C%B0%E9%9C%87%E5%BD%B1%E9%9F%BF-%E5%8F%B0%E7%A9%8D%E9%9B%BB%E5%BB%A0%E5%8D%80%E7%87%9F%E9%81%8B%E6%AD%A3%E5%B8%B8-051429015.html\\">竹科未受地震影響  台積電廠區營運正常</a>
(2020/09/30 中央社)

<a href=\\"https://tw.stock.yahoo.com/news/%E5%8F%B0%E7%A9%8D%E9%9B%BB%E5%90%AB%E9%8A%85%E5%BB%A2%E6%B6%B2%E5%86%8D%E7%94%9F%E6%8A%80%E8%A1%93-%E5%B9%B4%E5%89%B5%E9%80%BE%E5%84%84%E5%85%83%E5%9B%9E%E6%94%B6%E6%B8%9B%E5%BB%A2%E6%95%88%E7%9B%8A-045114599.html\\">台積電含銅廢液再生技術 年創逾億元回收減廢效益</a>
(2020/09/30 鉅亨網)

<a href=\\"https://tw.stock.yahoo.com/news/%E5%8F%B0%E8%82%A1%E7%9B%A4%E4%B8%AD-%E7%A7%8B%E7%AF%80%E9%95%B7%E5%81%87%E5%89%8D%E8%A7%80%E6%9C%9B-%E5%8F%B0%E8%82%A1%E9%87%8D%E5%9B%9E12500%E9%BB%9E%E4%B9%8B%E4%B8%8A-032027482.html\\">【台股盤中】秋節長假前觀望 台股重回12500點之上</a>
(2020/09/30 中央社)

<a href=\\"https://tw.stock.yahoo.com/news/%E5%8D%8A%E5%B0%8E%E9%AB%94-%E5%8F%B0%E7%A9%8D%E9%9B%BB%E5%90%AB%E9%8A%85%E5%BB%A2%E6%B6%B2%E5%86%8D%E7%94%9F%E6%8A%80%E8%A1%93-%E5%B9%B4%E5%89%B5%E9%80%BE%E5%84%84%E5%85%83%E7%B6%A0%E8%89%B2%E6%95%88%E7%9B%8A-022733402.html\\">《半導體》台積電含銅廢液再生技術 年創逾億元綠色效益</a>
(2020/09/30 時報資訊)

<a href=\\"https://tw.stock.yahoo.com/news/%E5%8F%B0%E7%A9%8D%E9%9B%BB%E5%BB%BA%E9%8A%85%E7%89%A9%E6%96%99%E5%86%8D%E7%94%9F%E5%BE%AA%E7%92%B0-%E5%B9%B4%E5%89%B5%E9%80%BE1%E5%84%84%E5%85%83%E7%B6%A0%E8%89%B2%E6%95%88%E7%9B%8A-014324143.html\\">台積電建銅物料再生循環 年創逾1億元綠色效益</a>
(2020/09/30 中央社)

<a href=\\"https://tw.stock.yahoo.com/news/%E5%90%84%E5%A0%B1%E8%A6%81%E8%81%9E-%E5%8F%B0%E7%A9%8D%E8%B5%B4%E7%BE%8E%E8%A8%AD%E5%BB%A0-ait%E6%8C%89%E8%AE%9A-233636405.html\\">《各報要聞》台積赴美設廠 AIT按讚</a>
(2020/09/30 時報資訊)

<a href=\\"https://tw.stock.yahoo.com/news/%E7%86%B1%E9%96%80%E6%97%8F%E7%BE%A4-%E5%8F%B0%E7%A9%8D%E9%9B%BB%E4%B9%8F%E4%BA%86-%E6%A6%82%E5%BF%B5%E8%82%A1%E5%90%84%E8%87%AA%E8%A1%A8%E6%85%8B-074320104.html\\">《熱門族群》台積電乏了 概念股各自表態</a>
(2020/09/29 時報資訊)
"
`;

exports[`lib/template getStockCaptionTextTemplate() it should return correct template 1`] = `"🔼 6443 元晶 24.85 | +2.25 (+9.96%)"`;

exports[`lib/template getStockHTMLTemplate() it should return correct template 1`] = `
"<pre>6443 元晶 | <b>24.85</b> | +2.25 (+9.96%)

總量 40534 | 內盤 8372 | 外盤 31223

最高 24.85 | 最低 22.2 | 開盤 - | 昨收 22.6

        委買        |        委賣        
----------------------------------------
      13218      市價                     
       5053   24.85                     
         60    24.8                     
         51   24.75                     
         66    24.7                     

</pre>
"
`;
